src/ton/TonTools.ts

```diff
@@ -0,0 +1,266 @@
+import {
+  EdgeCurrencyInfo,
+  EdgeCurrencyTools,
+  EdgeEncodeUri,
+  EdgeIo,
+  EdgeLog,
+  EdgeMetaToken,
+  EdgeParsedUri,
+  EdgeToken,
+  EdgeTokenMap,
+  EdgeWalletInfo,
+  JsonObject
+} from 'edge-core-js/types'
+import { base16 } from 'rfc4648'
+
+import { PluginEnvironment } from '../common/innerPlugin'
+import { asSafeCommonWalletInfo } from '../common/types'
+import { encodeUriCommon, parseUriCommon } from '../common/uriHelpers'
+import {
+  getLegacyDenomination,
+  mergeDeeply,
+  shuffleArray
+} from '../common/utils'
+import {
+  asTonPrivateKeys,
+  TonInfoPayload,
+  TonNetworkInfo,
+  wasTonPrivateKeys
+} from './tonTypes'
+
+const ONE_HOUR = 1000 * 60 * 60
+let clientsLastUpdate = 0
+
+export class TonTools implements EdgeCurrencyTools {
+  io: EdgeIo
+  builtinTokens: EdgeTokenMap
+  currencyInfo: EdgeCurrencyInfo
+  networkInfo: TonNetworkInfo
+  initOptions: JsonObject
+  log: EdgeLog
+
+  clients: TonClient[]
+  fetchAdaptor: ReturnType<typeof createFetchAdapter>
+
+  constructor(env: PluginEnvironment<TonNetworkInfo>) {
+    const { builtinTokens, currencyInfo, initOptions, io, networkInfo } = env
+    this.io = io
+    this.currencyInfo = currencyInfo
+    this.builtinTokens = builtinTokens
+    this.networkInfo = networkInfo
+    this.initOptions = initOptions
+    this.log = env.log
+
+    // Barebones fetch adaptor to work specifically with the @ton/ton library
+    this.fetchAdaptor = createFetchAdapter({
+      fetch: async (
+        input: RequestInfo | URL,
+        init?: RequestInit | undefined
+      ): Promise<Response> => {
+        if (!(input instanceof Request)) throw new Error('Invalid input')
+
+        const fetchHeaders: Record<string, string> = {}
+        for (const [key, value] of input.headers.entries()) {
+          fetchHeaders[key] = value
+        }
+
+        const res = await io.fetch(input.url, {
+          headers: fetchHeaders,
+          method: input.method,
+          body: await input.arrayBuffer()
+        })
+
+        const out = {
+          headers: res.headers,
+          ok: res.ok,
+          status: res.status,
+          arrayBuffer: async () => await res.arrayBuffer(),
+          json: async () => await res.json(),
+          text: async () => await res.text()
+        }
+
+        return out as Response
+      }
+    })
+
+    this.clients = [
+      this.networkInfo.tonCenterUrl,
+      ...this.networkInfo.defaultOrbUrls
+    ].map(url => {
+      return new TonClient({
+        endpoint: `${url}/jsonRPC`,
+        httpAdapter: this.fetchAdaptor
+      })
+    })
+  }
+
+  getClients(): TonClient[] {
+    if (Date.now() - clientsLastUpdate > ONE_HOUR) {
+      this.updateClients()
+        .then(() => {
+          clientsLastUpdate = Date.now()
+        })
+        .catch(this.log.error)
+    }
+    return shuffleArray(this.clients)
+  }
+
+  async updateClients(): Promise<void> {
+    const res = await this.io.fetchCors(this.networkInfo.orbsNetworkUrl)
```

Comment:
This seems like a job for the info server. The info server payload for TON should include an `orbsUrls` field, and then we should have a background process on the info server that fetches this information and shoves it in CouchDB. Then, the core info endpoint would merge the `orbsUrls` field into the payload on it way out the door.